### PR TITLE
Adjust taiko (again) to limit minimum aspect ratio to 5:4

### PR DESCRIPTION
**Pull request by peppy**
Anything lower than this looks stupid and broken. This should not affect any normal player. If it does, they should be using a "lane cover" mod instead. Or classic mod.

Why change this for a *third or fourth* time? Because to a normal user the current behaviour would look like a bug. This should be a middle ground which I hope everyone can accept.

### CODE REVIEW

**Comment by bdach** [comment]
The proportions look better, but the hitobjects appear to be going hyperspeed on very narrow windows:

https://user-images.githubusercontent.com/20418176/217940270-fda2184e-f6bb-4826-b8fc-a6c95f4d9878.mp4

I'm not sure if this is expected anymore? Logically I'd say it shouldn't be happening but I don't even know anymore with all this jankage.

**Comment by peppy** [comment]
That is indeed weird and I have no immediate idea why it would happen.

**Comment by bdach** [comment]
I was suspecting that it's due to something to do with https://github.com/ppy/osu/pull/22325/files. That said I was a bit too knackered yesterday to confirm that it's that and figure out what would be the proper thing to do there to not have this happen.

**Comment by peppy** [comment]
You're probably right. I thought what I had felt too simple for the amount of changes which were done over the last few releases. I'll take a look today regardless.

File: osu.Game.Rulesets.Taiko/UI/TaikoPlayfieldAdjustmentContainer.cs [hunk 1]
```diff
@@ -24,10 +25,17 @@ protected override void Update()
             // Players coming from stable expect to be able to change the aspect ratio regardless of the window size.
             // We originally wanted to limit this more, but there was considerable pushback from the community.
             //
-            // As a middle-ground, the aspect ratio can still be adjusted in the downwards direction but has a maximum limit.
+            // As a middle-ground, the aspect ratio can still be adjusted in the downwards direction but has a maximum limit (5:4 aspect).
             // This is still a bit weird, because readability changes with window size, but it is what it is.
-            if (LockPlayfieldMaxAspect.Value && Parent.ChildSize.X / Parent.ChildSize.Y > default_aspect)
-                height *= Math.Clamp(Parent.ChildSize.X / Parent.ChildSize.Y, 0.4f, 4) / default_aspect;
```
Comment (sw1tchbl4d3r) [inline osu.Game.Rulesets.Taiko/UI/TaikoPlayfieldAdjustmentContainer.cs:30]: Removing this clamp breaks things on wide aspect ratios, the playfield seemingly gets unloaded?

https://user-images.githubusercontent.com/54949148/218187388-a164064b-2388-4d60-b820-0179f2e6eef4.mp4

Comment (mk56-spn) [reply osu.Game.Rulesets.Taiko/UI/TaikoPlayfieldAdjustmentContainer.cs:30]: that looks cropped out by masking moreso than unloaded

Comment (bdach) [reply osu.Game.Rulesets.Taiko/UI/TaikoPlayfieldAdjustmentContainer.cs:30]: Correct, this is a masking/presence concern. Doesn't change the fact that it probably needs to be addressed somehow.

**Comment by peppy** [comment]
I'm not sure anymore. I tried applying the clamping back but the numbers no longer match. Also making the window short seems to make notes travel slower again expectations. This is cursed.

**Comment by sw1tchbl4d3r** [comment]
I've tried to solve this as well, and ended up with https://github.com/sw1tchbl4d3r/osu/commit/c82b74a5af40b0e9af71037263cc1b1edc1b439e this crude code right here, which does work, but I will be honest, I am not quite sure why.
This whole thing makes me feel like clamping the ratio to avoid the masking error is the wrong approach, as the code above has the same issue lazer has had for a pretty long time, that being the notes getting smaller and going hyperspeed on super wide aspect ratios due to the aspect ratio being clamped to 0.4 - 4 in this weird way...

**Comment by bdach** [comment]
I stared at this a bit today and yep, that clamp is pretty cursed. I may have something to offer as a fix, though - namely changing the layouting in `TaikoPlayfieldAdjustmentContainer` to... not move contents off screen, like so:

```diff
diff --git a/osu.Game.Rulesets.Taiko/UI/TaikoPlayfieldAdjustmentContainer.cs b/osu.Game.Rulesets.Taiko/UI/TaikoPlayfieldAdjustmentContainer.cs
index e97bb195d2..c7f8fbf5c8 100644
--- a/osu.Game.Rulesets.Taiko/UI/TaikoPlayfieldAdjustmentContainer.cs
+++ b/osu.Game.Rulesets.Taiko/UI/TaikoPlayfieldAdjustmentContainer.cs
@@ -1,6 +1,7 @@
 // Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+using System;
 using osu.Framework.Bindables;
 using osu.Framework.Graphics;
 using osu.Game.Rulesets.UI;
@@ -37,9 +38,12 @@ protected override void Update()
                     height *= currentAspect / MINIMUM_ASPECT;
             }
 
+            // Limit the maximum relative height of the playfield to one-third of available area to avoid it masking out on extreme resolutions.
+            height = Math.Min(height, 1f / 3f);
             Height = height;
 
-            // Position the taiko playfield exactly one playfield from the top of the screen.
+            // Position the taiko playfield exactly one playfield from the top of the screen, if there is enough space for it.
+            // Note that the relative height cannot exceed one-third - if that limit is hit, the playfield will be exactly centered.
             RelativePositionAxes = Axes.Y;
             Y = height;
         }
```

In action:

https://user-images.githubusercontent.com/20418176/218569613-9da0f22e-614c-421f-ac0c-617e239c3af3.mp4

Thoughts?

**Comment by sw1tchbl4d3r** [comment]
The only direct problem I can see in that video is more notes becoming visible on wider aspect ratios than 16:9 without the classic mod, though that's something to be fixed in the timerange calculation.

**Comment by peppy** [comment]
That sounds about right. Outside of the min/max aspect ratios the time range needs to be fixed rather than flexible. ie. how it was before *all of these recent changes were applied*.

**Comment by sw1tchbl4d3r** [comment]
I think I've found a pretty good solution to this which seems pretty logical, mind if I open a separate PR for it? (@peppy)

**Comment by peppy** [comment]
@sw1tchbl4d3r please do, yes. i haven't had a chance to revisit this and even when i do, i know it's going to take a lot of trial and error.

**Pull request closed without merging**